dace/transformation/dataflow/map_distribution.py

```diff
@@ -0,0 +1,940 @@
+                       dace.Memlet.simple(root_name, '0'))
+
+        from dace.libraries.mpi import Bcast, Scatter, Gather
+
+        inputs = set()
+        for src, _, _, _, m in graph.in_edges(map_entry):
+            if not isinstance(src, nodes.AccessNode):
+                raise NotImplementedError
+            desc = src.desc(sdfg)
+            if not isinstance(desc, (data.Scalar, data.Array)):
+                raise NotImplementedError
+            if list(desc.shape) != m.src_subset.size_exact():
+                # Second attempt
+                # TODO: We need a solution for symbols not matching
+                if str(list(desc.shape)) != str(m.src_subset.size_exact()):
+                    raise NotImplementedError
+            inputs.add(src)
+
+        for inp in inputs:
+            desc = inp.desc(sdfg)
+
+            if isinstance(desc, data.Scalar):
+                local_access = graph.add_access(inp.data)
+                bcast_node = Bcast('_Bcast_')
+                graph.add_edge(inp, None, bcast_node, '_inbuffer',
+                               dace.Memlet.from_array(inp.data, desc))
+                graph.add_edge(root_node, None, bcast_node, '_root',
+                               dace.Memlet.simple(root_name, '0'))
+                graph.add_edge(bcast_node, '_outbuffer', local_access, None,
+                               dace.Memlet.from_array(inp.data, desc))
+                for e in graph.edges_between(inp, map_entry):
+                    graph.add_edge(local_access, None, map_entry, e.dst_conn,
+                                   dace.Memlet.from_array(inp.data, desc))
+                    graph.remove_edge(e)
+                continue
+
+            if isinstance(desc, data.Array):
+
+                local_name, local_arr = sdfg.add_temp_transient(
+                    [(desc.total_size) // sz], dtype=desc.dtype,
+                    storage=desc.storage)
+                local_access = graph.add_access(local_name)
+                scatter_node = Scatter('_Scatter_')
+                graph.add_edge(inp, None, scatter_node, '_inbuffer',
+                               dace.Memlet.from_array(inp.data, desc))
+                graph.add_edge(root_node, None, scatter_node, '_root',
+                               dace.Memlet.simple(root_name, '0'))
+                graph.add_edge(scatter_node, '_outbuffer', local_access, None,
+                               dace.Memlet.from_array(local_name, local_arr))
+                for e in graph.edges_between(inp, map_entry):
+                    graph.add_edge(local_access, None, map_entry, e.dst_conn,
+                                   dace.Memlet.from_array(local_name, local_arr))
+                    graph.remove_edge(e)
+                for e in graph.out_edges(map_entry):
+                    if e.data.data == inp.data:
+                        e.data = dace.Memlet.simple(local_name, params[0])
+                continue
+
+            raise NotImplementedError
```

Comment:
I'd modify the code above to `if`, `elif`, and `else` in this case in order to avoid confusion.